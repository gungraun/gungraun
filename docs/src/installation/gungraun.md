# Gungraun

Gungraun is divided into the library `gungraun` and the benchmark runner
`gungraun-runner`.

## Installation of the library

To start with Gungraun, add the following to your `Cargo.toml` file:

```toml
[dev-dependencies]
gungraun = "0.16.1"
```

or run

```bash
cargo add --dev gungraun@0.16.1
```

## Installation of the benchmark runner

To be able to run the benchmarks you'll also need the `gungraun-runner`
binary installed somewhere in your `$PATH`. Otherwise, there is no need to
interact with `gungraun-runner` as it is just an implementation detail.

### From Source

```shell
cargo install --version 0.16.1 gungraun-runner
```

There's also the possibility to install the binary somewhere else and point the
`GUNGRAUN_RUNNER` environment variable to the absolute path of the
`gungraun-runner` binary like so:

```shell
cargo install --version 0.16.1 --root /tmp gungraun-runner
GUNGRAUN_RUNNER=/tmp/bin/gungraun-runner cargo bench --bench my-bench
```

### Binstall

The `gungraun-runner` binary is
[pre-built](https://github.com/gungraun/gungraun/releases/tag/v0.16.1)
for most platforms supported by valgrind and easily installable with
[binstall](https://github.com/cargo-bins/cargo-binstall)

```shell
cargo binstall gungraun-runner@0.16.1
```

## Updating

When updating the `gungraun` library, you'll also need to update
`gungraun-runner` and vice-versa or else the benchmark runner will exit
with an error.

### In the Github CI

Since the `gungraun-runner` version must match the `gungraun` library
version it's best to automate this step in the CI. A job step in the github
actions CI could look like this

```yaml
- name: Install gungraun-runner
  run: |
    version=$(cargo metadata --format-version=1 |\
      jq '.packages[] | select(.name == "gungraun").version' |\
      tr -d '"'
    )
    cargo install gungraun-runner --version $version
```

Or, speed up the overall installation time with `binstall` using the
[taiki-e/install-action](https://github.com/taiki-e/install-action)

```yaml
- uses: taiki-e/install-action@cargo-binstall
- name: Install gungraun-runner
  run: |
    version=$(cargo metadata --format-version=1 |\
      jq '.packages[] | select(.name == "gungraun").version' |\
      tr -d '"'
    )
    cargo binstall --no-confirm gungraun-runner --version $version
```
