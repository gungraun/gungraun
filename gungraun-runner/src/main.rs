//! The `gungraun-runner` binary
use std::io::Write;

use colored::{control, Colorize};
use env_logger::Env;
use gungraun_runner::error::Error;
use gungraun_runner::runner::envs;
use log::{error, warn};

/// The main function of the `gungraun-runner` binary
///
/// We initialize the logging interface and configure the usage of colors as early as possible here.
/// Then we're printing warnings with [`print_warnings`] and finally call the main
/// [`gungraun_runner::runner::run`] library function catching and printing
/// [`gungraun_runner::error::Error`]s.
fn main() {
    // Configure the colored crate to respect GUNGRAUN_COLOR and CARGO_TERM_COLOR
    let gungraun_color = std::env::var(envs::GUNGRAUN_COLOR).ok();
    if let Some(var) = gungraun_color
        .clone()
        .or_else(|| std::env::var(envs::CARGO_TERM_COLOR).ok())
    {
        if var == "never" {
            control::set_override(false);
        } else if var == "always" {
            control::set_override(true);
        } else {
            // do nothing
        }
    }

    // Configure the env_logger crate to respect GUNGRAUN_COLOR and CARGO_TERM_COLOR
    env_logger::Builder::from_env(
        Env::default()
            .filter_or(envs::GUNGRAUN_LOG, "warn")
            .write_style(
                gungraun_color.map_or_else(|| envs::CARGO_TERM_COLOR, |_| envs::GUNGRAUN_COLOR),
            ),
    )
    .format(|buf, record| {
        writeln!(
            buf,
            "{}: {:<5}: {}",
            record
                .module_path()
                .unwrap_or_else(|| record.module_path_static().unwrap_or("???")),
            match record.level() {
                log::Level::Error => "Error".red().bold(),
                log::Level::Warn => "Warn".yellow().bold(),
                log::Level::Info => "Info".green().bold(),
                log::Level::Debug => "Debug".blue().bold(),
                log::Level::Trace => "Trace".cyan().bold(),
            },
            record.args()
        )
    })
    .init();

    print_warnings();
    match gungraun_runner::runner::run() {
        Ok(()) => {}
        Err(error) => {
            if let Some(Error::RegressionError(is_fatal)) = error.downcast_ref::<Error>() {
                if *is_fatal {
                    error!("{error}");
                }
                std::process::exit(3)
            } else {
                error!("{error}");
                std::process::exit(1)
            }
        }
    }
}

/// Print warnings for deprecated usages of environment variables
fn print_warnings() {
    if std::env::var("IAI_ALLOW_ASLR").is_ok() {
        warn!("The IAI_ALLOW_ASLR environment variable changed to GUNGRAUN_ALLOW_ASLR");
    }

    if std::env::var("RUST_LOG").is_ok() {
        warn!("The RUST_LOG environment variable to set the log level changed to GUNGRAUN_LOG");
    }

    if std::env::var("IAI_CALLGRIND_REGRESSION").is_ok() {
        warn!(
            "With version 0.17.0, the name of the environment variable `IAI_CALLGRIND_REGRESSION` \
             has changed to `GUNGRAUN_CALLGRIND_LIMITS`."
        );
    }

    for var in [
        "ALLOW_ASLR",
        "BASELINE",
        "BBV_ARGS",
        "CACHEGRIND_ARGS",
        "CACHEGRIND_LIMITS",
        "CACHEGRIND_METRICS",
        "CALLGRIND_ARGS",
        "CALLGRIND_LIMITS",
        "CALLGRIND_METRICS",
        "COLOR",
        "DEFAULT_TOOL",
        "DHAT_ARGS",
        "DHAT_LIMITS",
        "DHAT_METRICS",
        "DRD_ARGS",
        "DRD_METRICS",
        "FILTER",
        "HELGRIND_ARGS",
        "HELGRIND_METRICS",
        "HOME",
        "LIST",
        "LOAD_BASELINE",
        "LOG",
        "MASSIF_ARGS",
        "MEMCHECK_ARGS",
        "MEMCHECK_METRICS",
        "NOCAPTURE",
        "NOSUMMARY",
        "OUTPUT_FORMAT",
        "REGRESSION_FAIL_FAST",
        "SAVE_BASELINE",
        "SAVE_SUMMARY",
        "SEPARATE_TARGETS",
        "SHOW_GRID",
        "SHOW_INTERMEDIATE",
        "SHOW_ONLY_COMPARISON",
        "TOLERANCE",
        "TOOLS",
        "TRUNCATE_DESCRIPTION",
        "VALGRIND_ARGS",
    ] {
        let old = format!("IAI_CALLGRIND_{var}");
        if std::env::var(&old).is_ok() {
            let new = format!("GUNGRAUN_{var}");
            if std::env::var(&new).is_err() {
                warn!(
                    "With version 0.17.0, the name of the environment variable `{old}` has \
                     changed to `{new}`."
                );
            }
        }
    }
}
