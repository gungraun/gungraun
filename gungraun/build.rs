//! The build script

// spell-checker: ignore rustified iquote iquotevalgrind

#[cfg(feature = "client_requests_defs")]
mod imp {
    use std::borrow::Cow;
    use std::fmt::Display;
    use std::io::{BufRead, BufReader, Cursor};
    use std::path::PathBuf;

    use bindgen::{builder, Bindings};
    use rustc_version::{version, Version};
    use strum::{EnumIter, IntoEnumIterator};

    #[derive(Debug)]
    struct Target {
        arch: String,
        env: String,
        os: String,
        triple: String,
        vendor: String,
    }

    #[derive(EnumIter, Debug, PartialEq, Eq)]
    enum Support {
        Arm,
        Aarch64,
        X86,
        X86_64,
        Riscv64,
        Native,
        No,
    }

    impl Display for Support {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            let support = format!("{self:?}").to_lowercase();
            f.write_str(&support)
        }
    }

    impl Target {
        fn from_env() -> Self {
            Self {
                arch: std::env::var("CARGO_CFG_TARGET_ARCH").unwrap(),
                env: std::env::var("CARGO_CFG_TARGET_ENV").unwrap(),
                os: std::env::var("CARGO_CFG_TARGET_OS").unwrap(),
                vendor: std::env::var("CARGO_CFG_TARGET_VENDOR").unwrap(),
                triple: std::env::var("TARGET").unwrap(),
            }
        }
    }

    fn print_client_requests_support(value: &Support) {
        println!("cargo:rustc-cfg=client_requests_support=\"{value}\"");
    }

    fn include_dirs(target: &Target) -> impl Iterator<Item = String> {
        [
            Cow::Owned(format!(
                "IAI_CALLGRIND_{}_VALGRIND_INCLUDE",
                target.triple.replace('-', "_").to_ascii_uppercase()
            )),
            Cow::Borrowed("IAI_CALLGRIND_VALGRIND_INCLUDE"),
        ]
        .into_iter()
        .filter_map(|env| std::env::var(env.as_ref()).ok())
    }

    fn build_native(target: &Target) {
        let mut builder = cc::Build::new();

        for env in include_dirs(target) {
            builder.include(env);
        }

        if target.os == "freebsd" {
            builder.include("/usr/local/include");
        }

        if let Ok(env) = std::env::var("GUNGRAUN_CROSS_TARGET") {
            let path = PathBuf::from("/valgrind/target/valgrind")
                .join(env)
                .join("include");
            builder.include(path);
        }

        builder.include("valgrind/include");

        builder
            .debug(true)
            .file("valgrind/native.c")
            .compile("native");
    }

    fn build_bindings(target: &Target) -> Bindings {
        let mut builder = builder();

        for env in include_dirs(target) {
            builder = builder.clang_arg(format!("-iquote{env}"));
        }

        if let Ok(env) = std::env::var("GUNGRAUN_CROSS_TARGET") {
            let path = PathBuf::from("/valgrind/target/valgrind")
                .join(env)
                .join("include");
            builder = builder.clang_arg(format!("-iquote{}", path.display()));
        }

        if target.os == "freebsd" {
            builder = builder.clang_arg("-iquote/usr/local/include");
        }

        let bindings = builder
            .clang_arg("-iquote/usr/include")
            .clang_arg("-iquotevalgrind/include")
            .header("valgrind/wrapper.h")
            .allowlist_var("IC_IS_PLATFORM_SUPPORTED_BY_VALGRIND")
            .allowlist_var("IC_VALGRIND_MAJOR")
            .allowlist_var("IC_VALGRIND_MINOR")
            .allowlist_type("IC_.*ClientRequest")
            .rustified_enum("IC_.*ClientRequest")
            .layout_tests(false)
            .parse_callbacks(Box::new(bindgen::CargoCallbacks::new()))
            .generate()
            .expect("Generating binding should succeed");

        let out_dir = std::env::var("OUT_DIR").map(PathBuf::from).unwrap();
        let path = out_dir.join("bindings.rs");
        bindings.write_to_file(path).unwrap();
        bindings
    }

    // Return the rust version if running rustc was successful
    fn get_rust_version() -> Option<Version> {
        version().ok()
    }

    pub fn main() {
        println!("cargo:rerun-if-changed=valgrind/wrapper.h");
        println!("cargo:rerun-if-changed=valgrind/native.c");
        println!("cargo:rerun-if-env-changed=IAI_CALLGRIND_VALGRIND_INCLUDE");
        println!("cargo:rerun-if-env-changed=GUNGRAUN_CROSS_TARGET");
        println!("cargo:rerun-if-env-changed=TARGET");

        // rustc-check-cfg is introduced in rust with version 1.80 and avoids the compiler warnings
        // in version >= 1.80.0. Printing it when compiling with versions < 1.80 triggers a warning,
        // too. To get the best of both worlds we check against the currently active rust version.
        if let Some(rust_version) = get_rust_version() {
            if rust_version.major >= 1 && rust_version.minor >= 80 {
                let values = Support::iter()
                    .map(|s| format!("\"{s}\""))
                    .collect::<Vec<String>>()
                    .join(",");
                println!("cargo:rustc-check-cfg=cfg(client_requests_support,values({values}))");
            }
        }

        let target = Target::from_env();

        // When building the docs on docs.rs we can take a shortcut
        if std::env::var("DOCS_RS").is_ok() {
            print_client_requests_support(&Support::X86_64);
            build_bindings(&target);
            build_native(&target);
            return;
        }

        let bindings = build_bindings(&target);

        // These guards mirror the checks in the `valgrind.h` header file
        let support = if target.arch == "x86_64"
            && (target.os == "linux"
                || target.os == "freebsd"
                || (target.vendor == "apple" && target.os == "darwin")
                || (target.os == "windows" && target.env == "gnu")
                || ((target.vendor == "sun") || target.vendor == "pc") && target.os == "solaris")
        {
            Some(Support::X86_64)
        } else if target.arch == "x86"
            && (target.os == "linux"
                || target.os == "freebsd"
                || (target.vendor == "apple" && target.os == "darwin")
                || (target.os == "windows" && target.env == "gnu")
                || ((target.vendor == "sun") || target.vendor == "pc") && target.os == "solaris")
        {
            Some(Support::X86)
        } else if target.arch == "arm" && target.os == "linux" && target.env == "gnu" {
            Some(Support::Arm)
        } else if target.arch == "aarch64"
            && (target.os == "freebsd" || (target.os == "linux" && target.env == "gnu"))
        {
            Some(Support::Aarch64)
        } else if target.arch == "riscv64gc" && target.os == "linux" {
            Some(Support::Riscv64)
        } else {
            let re = regex::Regex::new(
                r"IC_IS_PLATFORM_SUPPORTED_BY_VALGRIND.*?=\s*(?<value>true|false)",
            )
            .expect("Regex should compile");
            let reader = BufReader::new(Cursor::new(bindings.to_string()));
            let mut support = None;
            for line in reader.lines().map(Result::unwrap) {
                if let Some(caps) = re.captures(&line) {
                    let value = caps.name("value").unwrap().as_str();
                    if value == "false" {
                        support = Some(Support::No);
                    } else if value == "true" {
                        support = Some(Support::Native);
                    } else {
                        // do nothing
                    }
                    break;
                }
            }
            support
        };

        if let Some(support) = support {
            print_client_requests_support(&support);
            if support != Support::No {
                build_native(&target);
            }
        } else {
            eprintln!("{bindings}");
            panic!("Unable to set cfg value for client_requests_support");
        }
    }
}

#[cfg(not(feature = "client_requests_defs"))]
mod imp {
    pub fn main() {}
}

fn main() {
    imp::main();
}
