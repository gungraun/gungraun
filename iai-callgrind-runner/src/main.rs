//! The `iai-callgrind-runner` binary
use std::io::Write;

use colored::{control, Colorize};
use env_logger::Env;
use iai_callgrind_runner::error::Error;
use iai_callgrind_runner::runner::envs;
use log::{error, warn};

/// Print warnings for deprecated usages of environment variables
fn print_warnings() {
    if std::env::var("IAI_ALLOW_ASLR").is_ok() {
        warn!("The IAI_ALLOW_ASLR environment variable changed to IAI_CALLGRIND_ALLOW_ASLR");
    }

    if std::env::var("RUST_LOG").is_ok() {
        warn!(
            "The RUST_LOG environment variable to set the log level changed to IAI_CALLGRIND_LOG"
        );
    }
}

/// The main function of the `iai-callgrind-runner` binary
///
/// We initialize the logging interface and configure the usage of colors as early as possible here.
/// Then we're printing warnings with [`print_warnings`] and finally call the main
/// [`iai_callgrind_runner::runner::run`] library function catching and printing
/// [`iai_callgrind_runner::error::Error`]s.
fn main() {
    // Configure the colored crate to respect IAI_CALLGRIND_COLOR and CARGO_TERM_COLOR
    let iai_callgrind_color = std::env::var(envs::IAI_CALLGRIND_COLOR).ok();
    if let Some(var) = iai_callgrind_color
        .as_ref()
        .or(std::env::var(envs::CARGO_TERM_COLOR).ok().as_ref())
    {
        if var == "never" {
            control::set_override(false);
        } else if var == "always" {
            control::set_override(true);
        }
    }

    // Configure the env_logger crate to respect IAI_CALLGRIND_COLOR and CARGO_TERM_COLOR
    env_logger::Builder::from_env(
        Env::default()
            .filter_or(envs::IAI_CALLGRIND_LOG, "warn")
            .write_style(
                iai_callgrind_color
                    .map_or_else(|| envs::CARGO_TERM_COLOR, |_| envs::IAI_CALLGRIND_COLOR),
            ),
    )
    .format(|buf, record| {
        writeln!(
            buf,
            "{}: {:<5}: {}",
            record
                .module_path()
                .unwrap_or(record.module_path_static().unwrap_or("???")),
            match record.level() {
                log::Level::Error => "Error".red().bold(),
                log::Level::Warn => "Warn".yellow().bold(),
                log::Level::Info => "Info".green().bold(),
                log::Level::Debug => "Debug".blue().bold(),
                log::Level::Trace => "Trace".cyan().bold(),
            },
            record.args()
        )
    })
    .init();

    print_warnings();
    match iai_callgrind_runner::runner::run() {
        Ok(()) => {}
        Err(error) => match error.downcast_ref::<Error>() {
            Some(Error::IgnoredArgument(_)) => {
                warn!("{error}");
                std::process::exit(0)
            }
            Some(Error::RegressionError(false)) => std::process::exit(1),
            _ => {
                error!("{error}");
                std::process::exit(1)
            }
        },
    }
}
