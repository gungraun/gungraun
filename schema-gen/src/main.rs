//! A utility binary to create the json schema for the summary.json file
//!
//! This binary is not considered a part of the published `iai-callgrind-runner` package and is only
//! used during the development of `iai-callgrind`.
use std::fs::File;

use iai_callgrind_runner::runner::summary::BenchmarkSummary;
use schemars::generate::SchemaSettings;

fn main() {
    let generator = SchemaSettings::draft07().into_generator();
    serde_json::to_writer_pretty(
        File::create("summary.schema.json").unwrap(),
        &generator.into_root_schema_for::<BenchmarkSummary>(),
    )
    .expect("Schema creation should be successful");
}
