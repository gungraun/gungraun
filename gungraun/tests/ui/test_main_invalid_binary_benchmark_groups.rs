mod test_main_when_binary_benchmark_group_is_not_a_group {
    use gungraun::main;

    fn some_func() {}

    main!(binary_benchmark_groups = some_func);
}

mod test_main_when_config_is_not_a_binary_benchmark_config {
    use gungraun::{binary_benchmark_group, main};

    binary_benchmark_group!(
        name = some;
        benchmarks = |group: &mut BinaryBenchmarkGroup| {
            // do nothing
        }
    );

    main!(
        config = "some string";
        binary_benchmark_groups = some
    );
}

mod test_main_when_no_group {
    use gungraun::main;
    main!(binary_benchmark_groups = );
}

fn main() {}
