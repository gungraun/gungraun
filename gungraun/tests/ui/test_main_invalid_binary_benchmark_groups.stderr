error: The binary_benchmark_groups argument needs at least one `name` of a `binary_benchmark_group!`
  --> tests/ui/test_main_invalid_binary_benchmark_groups.rs:27:5
   |
27 |     main!(binary_benchmark_groups = );
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `main` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: use of undeclared crate or module `some_func`
 --> tests/ui/test_main_invalid_binary_benchmark_groups.rs:6:37
  |
6 |     main!(binary_benchmark_groups = some_func);
  |                                     ^^^^^^^^^ use of undeclared crate or module `some_func`

error[E0277]: the trait bound `InternalBinaryBenchmarkConfig: From<&str>` is not satisfied
  --> tests/ui/test_main_invalid_binary_benchmark_groups.rs:19:5
   |
19 | /     main!(
20 | |         config = "some string";
21 | |         binary_benchmark_groups = some
22 | |     );
   | |_____^ the trait `From<&str>` is not implemented for `InternalBinaryBenchmarkConfig`
   |
   = help: the following other types implement trait `From<T>`:
             <InternalBinaryBenchmarkConfig as From<BinaryBenchmarkConfig>>
             <InternalBinaryBenchmarkConfig as From<&BinaryBenchmarkConfig>>
             <InternalBinaryBenchmarkConfig as From<&mut BinaryBenchmarkConfig>>
   = note: required for `&str` to implement `Into<InternalBinaryBenchmarkConfig>`
   = note: this error originates in the macro `main` (in Nightly builds, run with -Z macro-backtrace for more info)
